### DISCUSSION TITLE
how to get the realtime updating relative pose of bodies?

### DISCUSSION THREAD

**Discussion by starimpact** [google-deepmind/mujoco#2229] [Asking for Help]
### Intro

Hi,

### My setup

```cpp
void getRelativePose(mjModel* m, int body_id, Vector3d& position, Quaterniond& orientation) {
    // get relative offset
    position = Vector3d(m->body_pos[3 * body_id], m->body_pos[3 * body_id + 1], m->body_pos[3 * body_id + 2]);

    // get relative pose
    orientation = Quaterniond(m->body_quat[4 * body_id + 0], m->body_quat[4 * body_id + 1], m->body_quat[4 * body_id + 2], m->body_quat[4 * body_id + 3]);
}
```

### My question

I use the upper code to get the relationship of bodies and their parents. But I found that they are static values. So, how can I get the realtime updating relative pose data when the robot is actuated ?

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by yuvaltassa** [comment]
Really, you searched the documentation? I don't think you did.

First, understand the difference between `mjModel` and `mjData`. Then the answer will become obvious.

Quoting from literally [the first section of the documentation](https://mujoco.readthedocs.io/en/latest/overview.html#key-features):

> MuJoCo separates simulation parameters into two data structures (C structs) at runtime:
> - [mjModel](https://mujoco.readthedocs.io/en/latest/APIreference/APItypes.html#mjmodel) contains the model description and is expected to remain constant. There are other structures embedded in it that contain simulation and visualization options, and those options need to be changed occasionally, but this is done by the user.
> - [mjData](https://mujoco.readthedocs.io/en/latest/APIreference/APItypes.html#mjdata) contains all dynamic variables and intermediate results. It is used as a scratch pad where all functions read their inputs and write their outputs – which then become the inputs to subsequent stages in the simulation pipeline. It also contains a preallocated and internally managed stack, so that the runtime module does not need to call memory allocation functions after the model is initialized.
